prometheusbp/promtest/spec.go

```diff
@@ -0,0 +1,80 @@
+package promtest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errLength         = errors.New("metric name should have a minimum of 3 parts, like prefix_name_suffix")
+	errCount          = errors.New("wrong metric count for prefix")
+	errPrometheusLint = errors.New("problem with Prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed conform to
+// the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	if err := validateSpec(metricPrefix, wantMetricCount); err != nil {
+		t.Error(err)
+	}
+}
+
+func validateSpec(metricPrefix string, wantMetricCount int) error {
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return err
+	}
+
+	var gotMetricCount int
+	for _, m := range metricFam {
+		metricName := *m.Name
+
+		if strings.Contains(metricName, metricPrefix) {
+			gotMetricCount++
+			if err := validatePromLint(metricName); err != nil {
+				return err
+			}
+			if err := validateName(metricName, metricPrefix); err != nil {
+				return err
+			}
```

Comment:
You don't want to return a single error here, as that will result in having to fix one bug at a time.  You want to return _all_ errors that are encountered.

see errorsbp.Batch